Add oidc-client, oidc-client-filter and oidc-token-propagation extensions

sberyozkin opened the pull request:
Fixes #5657

It does the following:

- adds `oidc-common` module which keeps the configuration common to both `quarkus-oidc` and `quarkus-oidc-client`/ - both need the same connection properties, auto-discovery support, client authentication, etc.

**OidcClient**

- adds `quarkus-oidc-client`: `Uni` - reactive, `Mutiny Vert.x WebClient` based `OidcClient` which currently supports `client-credentials` and `password` grants
- `OidcClient` authenticates with `client_secret_basic`, `client_secret_post`, `client_secret_jwt` (jwt private key to be added next)   
- `OidcClient` is injectable and returns `Tokens` which holds Access and Refresh tokens which can also be refreshed
- `Tokens` can be injected as well - the producer will use `OidcClient` internally to acquire/refresh the tokens - this feature is mainly for those applications which do not want to use `quarkus-oidc-rest-client` for some reasons
- `OidcClients` is an injectable container which contains a default `OidcClient` as well as the named clients. It can also be used to create `OidcClient` on demand 
- `OidcClientConfig` has a Map to keep various grant properties based on the convention that a grant name acts as a key; I've tried to avoid introducing grant-specific property holders to keep an option for the custom/non-standard grants open

**OidcClientRequestFilter**

- adds `quarkus-oidc-client-filter` - it provides a JAX-RS `OidcClientRequestFilter` which sets the access token acquired or refreshed by `OidcClient` as an HTTP `Authorization` Bearer scheme value.
- `OidcClientRequestFilter` can be added to all MP RestClients, or selectively, by using MP REST Client `@RegisterProvider(OidcClientRequestFilter.class)` or only adding `@OidcClientFilter` annotation

**AccessTokenRequestFilter**

- adds `quarkus-oidc-token-propagation` - it provides a JAX-RS `AccessTokenRequestFilter` which sets the current Bearer or Authorization Code Flow access token as an HTTP `Authorization` Bearer scheme value.
- `AccessTokenRequestFilter` can be added to all MP RestClients, or selectively, by using MP REST Client `@RegisterProvider(AccessTokenRequestFilter.class)` or only adding `@AccessToken` annotation


- `quarkus-rest-client` has been updated to support using annotations like `@OidcClientFilter` to add an associated provider. `quarkus-rest-client` will need another simple iteration to support more than one custom annotation but it is beyond the scope of this PR which is becoming quite large already

- tests, including the dev mode one
- initial documentation; some kind of quickstart will follow later once these new extensions have settled a bit 

CC @pedroigor @stuartwdouglas @gsmet @radcortez

sberyozkin committed:
Create oidc-client and oidc-rest-client extensions

jtama-op commented:
If I can help I'd be happy to. I think it's a really missing for the moment !

sberyozkin commented:
@jtama-op Thanks; I'm going to open it for review in a week or two max - early testing will be appreciated

pedroigor commented on docs/src/main/asciidoc/security-openid-connect-client.adoc:
```diff
@@ -0,0 +1,218 @@
+
+`quarkus-oidc-rest-client` extension provides `io.quarkus.oidc.rest.client.OidcClientRequestFilter` JAX-RS ClientRequestFilter which uses `OidcClient` to acquire the access token, refresh it if needed, and set it as an HTTP `Authorization` `Bearer` scheme value.
+
+By default this filter will get `OidcClient` to acquire the first pair of access and refresh tokens at its initialization time. If the access tokens are short lived and refresh tokens are not available then the token acquisition should be delayed with `quarkus.oidc-client.early-tokens-acquisition=false`.
```
I'm not sure it makes sense to set up the client with `user` because I'm not sure if that is a valid use case for chaining services. So, refresh tokens should only make sense if the application is a `web-app` that authenticates using code flow and needs to access downstream services.

For service-to-service communication it would makes more sense client credentials where refresh tokens are not supposed to be issued (in Keycloak we are changing this).

Btw, is it possible to use this filter using a bearer token sent to a `service` type application ? One thing nice here would be the possibility to automatically exchange tokens depending on the service the application is calling. So that another token is issued by the AS and meant to use in the target service.

sberyozkin replied:
Hi @pedroigor yeah, I'm just trying to position this client as a generic oidc client, beyond chaining the services or a service to service, you know we have many places were we use RestAssured or Keycloak admin api to get a token with a password grant; and I'm also keeping an option open for supporting the custom grants.
Right now this filter is not linked to the incoming tokens, the case where the coming in bearer token is used in the next invocation is not covered - in this simplest form there is nothing we can suggest but using the context propagation where `JsonWebToken` is injected in the client filter where one would just add its value to the Authorization header - I think I may add this simple filter anyway - and we've had an exchange tokens issue closed recently till better times :-) - so after this PR, I'll reopen that issue and we can enhance it with the exchange protocol, that should be interesting

sberyozkin replied:
Hey @pedroigor Re that new filter which would propagate the token, not sure I'll have time to put it into this PR, I'll see, but I'm thinking of naming it `TokenPropagationRequestFilter` as opposed to `OidcClientRequestFilter` which is shipped in this PR - the latter is not related in anyway to the endpoint authentication, while the former would pick up the injected token - in the 1st iteration it would just put it into the header, later - we will wire in the exchange protocol.
Does it sound good to you ?

pedroigor replied:
Sure

pedroigor commented on docs/src/main/asciidoc/security-openid-connect-client.adoc:
```diff
@@ -0,0 +1,218 @@
+
+== OidcClients
+
+`io.quarkus.oidc.client.OidcClient` is a container of `OidcClient`s - it includes a default `OidcClient` (which can also be injected directrly as described above) and named clients which can be configured like this:
```
Small typo in `directly`.

pedroigor commented on docs/src/main/asciidoc/security-openid-connect-client.adoc:
```diff
@@ -0,0 +1,218 @@
+
+    @GET
+    public String getResponse() {
+        OidcClient client = clients.getClient("jwt-secret");
```
Maybe this would be nicer if you provide a `@OidcClient` annotation where you can define a name and have the named  client injected.

sberyozkin replied:
Nice idea, I'll have a look; I'll keep `OidcClients` as I've already prototyped creating the clients on demand by copying/pasting your code from quarkus-oidc with a few cosmetic updates :-)

sberyozkin replied:
@pedroigor Hi Pedro, Martin @mkouba explained it is not going to work for ApplicationScoped endpoints, so for now I'll have to resolve this comment without any action. Using `OidcClients` is a bit more verbose to get a named client but I suppose it is a more complete approach - as it also allows to create the clients dynamically; As I suggested in the docs, if the users have multiple named clients then, at least when they have OIDC multi-tenancy, then can use a tenant id for a client look up, this selection of the named clients is supposed to be dynamic I guess; thanks, nice idea though :-)

pedroigor commented on extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/GrantTokens.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Access and Refresh tokens returned from a token grant request
+ */
+public class GrantTokens {
```
You probably have a reason for that ... Wouldn't be better `Tokens` ?

sberyozkin replied:
I did not like `GrantTokens` first but then somehow got used to it :-), no particular reason, just wanted to say the tokens is product of the grant request, but just `Tokens` is simpler I guess

sberyozkin replied:
done

pedroigor commented on extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/OidcClient.java:
```diff
@@ -0,0 +1,21 @@
+    /**
+     * Refreshes the grant tokens
+     */
+    Uni<GrantTokens> refreshTokens(String refreshToken);
```
Maybe a no-arg `refreshTokens` would help to avoid boilerplate code to get refresh token and then call this method?

sberyozkin replied:
But `OidcClientImpl` is not keeping the tokens itself. I guess in a good number of cases the users won't use this client directly but via the filter which is hiding these expiry and refresh tokens checks

pedroigor commented:
@sberyozkin Really nice stuff! Only a few comments, hope they make sense.

stuartwdouglas commented on extensions/oidc-rest-client/deployment/src/test/java/io/quarkus/oidc/rest/client/KeycloakRealmResourceManager.java:
```diff
@@ -0,0 +1,141 @@
+import io.quarkus.test.common.QuarkusTestResourceLifecycleManager;
+import io.restassured.RestAssured;
+
+public class KeycloakRealmResourceManager implements QuarkusTestResourceLifecycleManager {
```
At some point we should look at consolidating these into some kind of common utility. It could actually be useful for end users as well to have an easy way to setup keycloak for tests.

sberyozkin replied:
@stuartwdouglas +1, @cemnura is helping out with improving the test support for OIDC so we can work with him on it

sberyozkin replied:
@stuartwdouglas see https://github.com/quarkusio/quarkus/issues/13957

stuartwdouglas commented on docs/src/main/asciidoc/security-openid-connect-client.adoc:
```diff
@@ -0,0 +1,218 @@
+public class OidcClientRequestCustomFilter implements ClientRequestFilter {
+
+    @Inject
+    GrantTokens grantTokens;
```
I don't think you should allow direct injection of the GrantTokens, as there is no CDI scope that this maps to. It would be very easy to accidentally inject them into a long lived object, which will appear to work for a while until the tokens expire (e.g. likely work in tests and staging, then fail once the production env has been up for a while). This example only works because you have made the filter @RequestScoped, which is not something we really encourage.

I think you should instead inject a GrantTokensProducer, which as a get() and getAsync() method. This object can be application scoped and actually manage the tokens in a thread safe manner so that new tokens don't need to be acquired per request.

sberyozkin replied:
@stuartwdouglas As I mentioned above I have a working test - but I'll review that test, may be it is broken :-)

sberyozkin replied:
@stuartwdouglas Hi Stuart, I've resolved 2 related comments, thanks for highlighting the problems with this specific producer; so, first of all `GrantTokens` and `GrantTokensProducer` got renamed based on @pedroigor 's suggestion to simplify so the `Grant` part has gone from their names.
`TokensProducer` is now `@Singleton` but it `@RequestScoped Produces` the tokens. Because it is a singleton it gets the initial set of tokens on its initialization time only once and then it will either return them or refresh them. This should be correct now.
The other thing about `TokensProducer` and the `injected Tokens` is that it is not a main thing this PR is after - `OidcClient` already provides the direct async access to `Tokens`, and `OidcClientRequestFilter` manages them too which is what I'd expect most users would use, `injected Tokens` is a little utility for those users who just don't want to use the provided filter or `OidcClient` request directly and it now works correctly I believe.

By the way, I've added a check log code (where the log is checked that it contains a refresh token confirmation) but what I've found is that apparently, when multiple tests run, as in `oidc-client/deployment`, the `target` dir seems to be locked somehow, and this check log only works if I run the individual test,
i.e, the following works:
```
mvn install -Pdocker-keycloak -Ptest-keycloak -Dtest=OidcClientUserPasswordCustomFilterTestCase
```
This test would successfully detect `target/quarkus.log` and check the records there.
The same test fails with:
```
mvn install -Pdocker-keycloak -Ptest-keycloak
```
It runs more than one test, not only `OidcClientUserPasswordCustomFilterTestCase`, and `OidcClientUserPasswordCustomFilterTestCase` fails with the message that `target/quarkus.log` is not available.

Do you know what may be going on ? I've removed the checkLog code from it for now, as `integration-tests/oidc-client` also checks the log and it works fine

sberyozkin replied:
@stuartwdouglas I've also updated `integration-tests/oidc-client` to check that log only contains a single acquisition and refresh entry, there are a few tests there which depend on a singleton `OidcClientRequestFilter`, before I added `@Singleton` to it it re-acquiring for every test so yeah now it is all good I think

sberyozkin commented on extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/GrantTokensHelper.java:
```diff
@@ -0,0 +1,37 @@
+
+        if (currentTokens.isAccessTokenExpired()) {
+            // Particularly, a client_credentials response is not guaranteed to return a refresh token
+            Uni<GrantTokens> newTokensUni = currentTokens.getRefreshToken() != null
```
@stuartwdouglas yes there is a (very) small window there for the concurrent refresh or token acquisition but I thought it was not critical, this operation in itself can be repeated, OIDC servers will control a number of times the tokens will be refreshed but a a few extra refresh operations won't affect it. 
As opposed to having to sync, or did you mean implementing  `you need to track if there is an operation in progress and piggy back on the existing operation` in terms of `Uni` ? It sounds right, just need to figure how to do it :-)

sberyozkin replied:
@stuartwdouglas I've added some code there to make sure no unnecessary refresh or a postponed token acquisition occurs, have a look please, that should be fine.

sberyozkin commented on extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/GrantTokensProducer.java:
```diff
@@ -0,0 +1,15 @@
+import io.quarkus.oidc.client.GrantTokens;
+
+@RequestScoped
+public class GrantTokensProducer extends AbstractGrantTokensProducer {
```
This producer gets the initial set of tokens on the initialization time, and it needs to be called on every request for this producer to check if the token has expired

sberyozkin commented on extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/GrantTokensProducer.java:
```diff
@@ -0,0 +1,15 @@
+public class GrantTokensProducer extends AbstractGrantTokensProducer {
+
+    @Produces
+    public GrantTokens produceTokens() {
```
@stuartwdouglas I'll check, sure, I have a test checking the logs which confirm the tokens have been refreshed (with the filter which uses `@OidcClient`) and I have a test which verifies the tokens have been refreshed with `@GrantTokensProducer/@GrantTokens` but I'll add the log check there as well :-)

stuartwdouglas replied:
Your test works because you have explicitly marker the filter as @RequestScoped. But this means that every request will require a request to KC to get the keys. 

I think the keys should be stored in a singleton, which will mean that most requests don't need to make any requests to KC. When they keys expire the singleton can manage the refresh process and make sure they are only refreshed once.

sberyozkin replied:
@stuartwdouglas Oh, do you mean the filter is re-initialized on every request ? that is not good then yeah

sberyozkin commented on extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/OidcClientRecorder.java:
```diff
@@ -0,0 +1,194 @@
+                JsonObject json = resp.bodyAsJsonObject();
+                return json.getString("token_endpoint");
+            } else {
+                LOG.tracef("Discovery has failed, status code: %d", resp.statusCode());
```
@stuartwdouglas Likely not as it is not a connection error, but the discovery document retrieval problem

sberyozkin commented on extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/OidcClientRecorder.java:
```diff
@@ -0,0 +1,194 @@
+        grantParams.add("grant_type", grantType);
+        Credentials creds = oidcConfig.getCredentials();
+        if (OidcCommonUtils.isClientSecretPostAuthRequired(creds)) {
+            grantParams.add("client_id", oidcConfig.clientId.get());
```
@stuartwdouglas good point, now that this PR introduces `oidc-common` we can have `OIDCConstants` there, I'll see how it goes, if I have time I'll do it with this PR otheriwise I'll create a follow up issue and take care of it there

sberyozkin replied:
@stuartwdouglas Just did it, definitely worth it :-), moved a good number of those magic constants from `quarkus-oidc-client` and a few from `quarkus-oidc`, more common ones will be added there from now on

sberyozkin commented:
@stuartwdouglas @pedro
I've updated the issue description, I've added one more light-weight extension which ships a `TokenCredentialRequestFilter` (Pedro - slight change in the filter name :-)).
So as a summary, we have:
- `io.quarkus.oidc.client.OidcClient` (`quarkus-oidc-client` extension) which can be used to acquire or refresh the tokens directly or via the `Tokens` producer - independently of Mp RestClient. 
- `io.quarkus.oidc.client.filter.OidcClientRequestFilter` (`quarkus-oidc-client-filter` extension) - it makes it trivial to use `OidcClient` in the selected MP RestClients with the help of the `@OidcClientFilter` annotation

Both `OidcClient` and `OidcClientRequestFilter` are not related to the current endpoint's authentication requirements.

- `io.quarkus.oidc.token.propagation.TokenCredentialRequestFilter` (`quarkus-oidc-token-propagation` extension) - it makes it trivial to propagate the current Bearer or Authorization Code Flow access token in the selected MP RestClients with the help of the `@TokenCredential` annotation.  This extension does not know what `OidcClient` is and it expects that the endpoint must have been protected with `quarkus-oidc`.

I'm actually getting quite happy now with how it lays out. It is a large PR and it can't be perfect in the first iteration but IMHO it is quite ready to go in. We can tune some things between 1.11.0.CR1 and 1.11.0.Final.
Have another look if you can spot something that needs an immediate fix
thanks

stuartwdouglas commented on extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/OidcClientImpl.java:
```diff
@@ -0,0 +1,150 @@
+                JsonObject claims = new JsonObject(new String(Base64.getUrlDecoder().decode(parts[1]), StandardCharsets.UTF_8));
+                return claims.getLong(Claims.exp.name());
+            } catch (IllegalArgumentException ex) {
+                return null;
```
This still needs to be addressed.

sberyozkin replied:
@stuartwdouglas sorry, missed it. OK, a token grant response should include an expiry time, but if it is not included, (as in case of Keycloak) then an attempt is made to decode the access token as JWT token and get the exp claim - this exception is only to handle the `JsonObject.getLong` exception - should not ever happen in practice - but indeed, some debug message should be added

stuartwdouglas commented on extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/TokensHelper.java:
```diff
@@ -0,0 +1,76 @@
+public class TokensHelper {
+
+    private volatile Tokens tokens;
+    private ReentrantLock tokenAcquisitionLock = new ReentrantLock();
```
We can't really do this via locks when you are dealing with reactive code. getTokens() can be called by the same thread for reactive code, but for different requests. The logic below also has a few issues.

sberyozkin replied:
@stuartwdouglas, thanks a lot; I have to admit this possibility did not occur to me :-) - I was primarily focused on running this code from the client filter which at this moment of time is not reactive ready yet as if I understand it correctly - Steph said he was only planning to support the reactive filters in 'regular' RestEasy - but you are right it has to be done right - as I'd like to have these filters running eventually in the reactive mode :-)
Also, thanks for getting this code right, let me learn what has changed, cheers

sberyozkin replied:
@stuartwdouglas - compare and swap :-) - I'm getting inspired here and I've just pulled Java Concurrentcy in Practice 2nd edition from the bookshelf though I suspect it may've gone a bit outdated :-)

sberyozkin replied:
@stuartwdouglas - actually - I'm finally getting how to make sure `Uni` can be kept as a state property, it must be `deferred`, I've been thinking of keeping it in `quarkus-oidc` somewhere as a state value but could not get how to do it :-)

stuartwdouglas commented:
I have pushed a commit that updates the state management for the tokens.

sberyozkin commented:
@stuartwdouglas added some log statements as well

sberyozkin commented:
@stuartwdouglas added some log statements as well; I thought I left some clarification above but looks like I lost it, thanks - I missed this returning `null` without any info - in many cases the expiry value will be returned in the token grant response - but in case of KC it is encoded directly in the access token - so there we try to decode the token as JWT and if it works then get the long exp claim - I've split that code and added some debug statements; thanks

sberyozkin commented:
@pedroigor FYI, I've re-enabled JWT private key authentication test, I had to set both public and private key as the KC model client representation properties and it took me awhile to came to that solution :-) but the good news is we now know how to do it. (the key which was you saw as null earlier was caused by a corrupt key resource - I was copying some content around and somehow broke it)

sberyozkin commented:
Renamed `TokenCredential` annotation with a simpler `AccessToken` one in the `quarkus-oidc-token-propagation`

sberyozkin commented:
One more JWT private key authentication test and the `OidcClient Authentication` doc section have been added.
The first test was using the PEM file (providers like Google may prefer it), the new test uses the keystore (in JKS format for a start)

sberyozkin commented:
Hi @stuartwdouglas I'm going to go ahead with the merge now as I'd really like for this PR to make it to CR1; ping me please - or comment here if you think something else may need to be addressed, thanks
